delivery_postlogistics/postlogistics/web_service.py

```diff
@@ -0,0 +1,506 @@
+from io import BytesIO
+
+from odoo import _, exceptions
+
+import requests
+from PIL import Image
+
+_logger = logging.getLogger(__name__)
+
+_compile_itemid = re.compile(r"[^0-9A-Za-z+\-_]")
+_compile_itemnum = re.compile(r"[^0-9]")
+AUTH_PATH = "/WEDECOAuth/token"
+GENERATE_LABEL_PATH = "/api/barcode/v1/generateAddressLabel"
+
+
+class PostlogisticsWebService(object):
+
+    """ Connector with PostLogistics for labels using post.ch API
+
+    Handbook available here:
+    https://developer.post.ch/en/digital-commerce-api
+    https://wedec.post.ch/doc/swagger/index.html?
+        url=https://wedec.post.ch/doc/api/barcode/v1/swagger.yaml
+        #/Barcode/generateAddressLabel
+
+    Allows to generate labels
+
+    """
+
+    access_token = False
+    access_token_expiry = False
+    _lock = threading.Lock()
+
+    def __init__(self, company):
+        self.default_lang = company.partner_id.lang or "en"
+
+    def _get_language(self, lang):
+        """ Return a language to iso format from odoo format.
+
+        `iso_code` field in res.lang is not mandatory thus not always set.
+        Use partner language if available, otherwise use english
+
+        :param partner: partner browse record
+        :return: language code to use.
+
+        """
+        if not lang:
+            lang = self.default_lang
+        available_languages = ["de", "en", "fr", "it"]  # Given by API doc
+        lang_code = lang.split("_")[0]
+        if lang_code in available_languages:
+            return lang_code
+        return "en"
+
+    def _prepare_recipient(self, picking):
+        """ Create a ns0:Recipient as a dict from a partner
+
+        :param partner: partner browse record
+        :return a dict containing data for ns0:Recipient
+
+        """
+        partner = picking.partner_id
+
+        partner_name = partner.name or partner.parent_id.name
+        recipient = {
+            "name1": partner_name,
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
+            "email": partner.email or None,
+        }
+
+        if partner.country_id.code:
+            recipient["country"] = partner.country_id.code.upper()
+
+        if partner.street2:
+            recipient["addressSuffix"] = partner.street2
+
+        company_partner_name = partner.commercial_company_name
+        if company_partner_name and company_partner_name != partner_name:
+            recipient["name2"] = partner.parent_id.name
+            recipient["personallyAddressed"] = False
+
+        # Phone and / or mobile should only be displayed if instruction to
+        # Notify delivery by telephone is set
+
+        picking_packagings = picking._get_picking_postlogistic_packages().mapped(
+            "packaging_id"
+        )
+        is_phone_required = [
+            packaging
+            for packaging in picking_packagings
+            if packaging.postlogistics_option_code == "ZAW3213"
```

Comment:
This option should be defined on the partner, not on the packaging.

Reply:
well noted, to be updated

Reply:
Fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555